deepseg.cc

```diff
@@ -371,20 +379,35 @@ void calc_mask(calcinfo_t &info, timinginfo_t &ti) {
 			fprintf(stderr, "Unknown model type\n");
 			break;
 	}
-	ti.maskns=timestamp();
+	info.pti->ai_postns = diffnanosecs(timestamp(), t0);
+	t0 = timestamp();
 
 	// scale up into full-sized mask
-	cv::Mat tmpbuf;
-	cv::resize(info.ofinal,tmpbuf,cv::Size(info.raw.rows/info.ratio,info.raw.rows));
+	{
+		cv::Mat tmpbuf;
+		std::lock_guard<std::mutex> hold(info.lock_mask);
+		cv::resize(info.ofinal,tmpbuf,cv::Size(info.raw_current->rows/info.ratio,info.raw_current->rows));
+		// blur at full size for maximum smoothness
+		cv::blur(tmpbuf,info.mroi,info.blur);
+	}
 
-	// blur at full size for maximum smoothness
-	cv::blur(tmpbuf,info.mroi,info.blur);
+	info.pti->ai_maskns = diffnanosecs(timestamp(), t0);
+	info.pti->ai_totalns = diffnanosecs(timestamp(), t);
 }
 
 bool is_number(const std::string &s) {
   return !s.empty() && std::all_of(s.begin(), s.end(), ::isdigit);
 }
 
+void *calc_mask_thread(void *args) {
```

Comment:
What about having this as a proper `thread`-derived class?

Reply:
Do you mean something like:
```c++
class CalcMask : public std::thread
```
I'm not so familiar with the threading module. 
Is this the recommended usage?

Reply:
Reading up on the [docs](https://en.cppreference.com/w/cpp/thread/thread) directly inheriting may cause some issues with initialization order (due to the thread being possibly running before the constructor completes). The following way should be more stable against that scenario:

```cpp
class CalcMask final {
protected:
    enum class thread_state { INIT, RUNNING, DONE };
    volatile thread_state state;
    // All mutexes need to come before the thread
    std::thread thread; // Must be after state and all relevant mutexes, because of initialization order)
    void run() {
        while(thread_state::INIT == this->state)
            ; // Wait for constructor to complete initialization
        while(thread_state::RUNNING == this->state) {
            /* actual handling */
        }
    }

public:
    CalcMask() :
        state{thread_state::INIT},
        thread{this->run} {
        // Do all other initialization …
        state = thread_state::RUNNING;
    }

    ~CalcMask() {
        state = thread_state::DONE;
        thread.join();
    }
};
```

Reply:
I'm not sure, where the new `CalcMask` class will bring us. 
From a naive perspective, this looks unnecessary complex to me.
Where do you see the benefit of having the thread in a separate class?
What do you see inside the thread class? Only the `calc_mask_thread` method, or the whole mask calculation?

Reply:
The mask calculation + mutexes + thread-local data + functions to interact/synchronize with it.
In the first iteration mostly just the `calc_mask_thread` function + mutexes.

Reply:
If I understand your correctly, you want to move the whole mask calculation into a class?
At the moment the `calcinfo_t` struct is a container for data. Isn't creating a `CalcMask` class the same as extending the c like struct into a c++ class?
For me this looks like a larger restructure, including creating a .cpp and .h file for the new class. 
Do you think we should do this as part of this PR?

On the other hand, I'm not so sure if having a class containing just the thread context makes things nicer than the current implementation. Having the thread context, mutex and run loop at one place, but the function called by the run loop and the data associated with the mutex at a different place doesn't sound ideal to me.

What do you think?

Reply:
There's a subtle difference between those two structures which mostly affects how things can be interacted with from code that is not C++. See #87 for issues and challenges with that one.

So the basic idea should be to keep the publicly visible part of the structure as simple as possible, and especially try to avoid C++-isms there, while the internal `CalcMask` object can be as complicated as we need it to be. Think of it kinda as `calcinfo_t` as the parameters and `CalcMask` as the runtime data.

Reply:
I see the point, that you want to have the lib as pure C.
As we now have the split, I would see the threading only as part of the app.

Just for me to get a better overall picture, what is your plan for the app? 
Do you want to restructure the whole main function into a class?
Will this be all part of the above mentioned `CalcMask` class, or will there be more then one class?

Reply:
The `CalcMask` class is likely only one of many classes, centered around the aspect of getting the segmentation mask from an input frame. Other classes/modules may perform tasks like frame rate decoupling (re-use mask for multiple frames, until next mask has been calculated) or image post-processing (hologram effects, light boxing, …).

Reply:
Thank you for the explanation. It's easier for me, if I can see the big picture.

Reply:
Looks resolved :smile: